File: hphp/hack/src/hhbc/emit_fatal.ml
(**
 * Copyright (c) 2017, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the "hack" directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 *
*)

open Instruction_sequence
open Hhbc_ast

exception IncludeTimeFatalException of FatalOp.t * Pos.t * string

let raise_fatal_runtime p message =
  raise (IncludeTimeFatalException (FatalOp.Runtime, p, message))

let raise_fatal_parse p message =
  raise (IncludeTimeFatalException (FatalOp.Parse, p, message))

let emit_fatal op pos message =
  gather
  [
    Emit_pos.emit_pos pos;
    instr_string message;
    instr (IOp (Fatal op))
  ]

let emit_fatal_runtime pos message =
  emit_fatal FatalOp.Runtime pos message
let emit_fatal_runtimeomitframe pos message =
  emit_fatal FatalOp.RuntimeOmitFrame pos message

let emit_fatal_for_break_continue pos level =
  let suffix = if level = 1 then "" else "s" in
  let message =
    Printf.sprintf "Cannot break/continue %d level%s" level suffix
  in
  emit_fatal_runtime pos message

File: hphp/hack/src/hhbc/emit_fatal.rs
// Copyright (c) Facebook, Inc. and its affiliates.
//
// This source code is licensed under the MIT license found in the
// LICENSE file in the "hack" directory of this source tree.

use emit_pos_rust::emit_pos;
use hhbc_ast_rust::FatalOp;
use instruction_sequence::{instr, Error, InstrSeq};
use oxidized::pos::Pos;

pub fn raise_fatal_runtime(pos: &Pos, msg: impl Into<String>) -> Error {
    Error::IncludeTimeFatalException(FatalOp::Runtime, pos.clone(), msg.into())
}

pub fn raise_fatal_parse(pos: &Pos, msg: impl Into<String>) -> Error {
    Error::IncludeTimeFatalException(FatalOp::Parse, pos.clone(), msg.into())
}

pub fn emit_fatal(op: FatalOp, pos: &Pos, msg: impl AsRef<str>) -> InstrSeq {
    InstrSeq::gather(vec![
        emit_pos(pos),
        instr::string(msg.as_ref()),
        instr::fatal(op),
    ])
}

pub fn emit_fatal_runtime(pos: &Pos, msg: impl AsRef<str>) -> InstrSeq {
    emit_fatal(FatalOp::Runtime, pos, msg)
}

pub fn emit_fatal_runtimeomitframe(pos: &Pos, msg: impl AsRef<str>) -> InstrSeq {
    emit_fatal(FatalOp::RuntimeOmitFrame, pos, msg)
}

pub fn emit_fatal_for_break_continue(pos: &Pos, level: usize) -> InstrSeq {
    let suffix = if level == 1 { "" } else { "s" };
    let msg = format!("Cannot break/continue {} level{}", level, suffix);
    emit_fatal_runtime(pos, msg)
}
